types/src/eth.rs

```diff
@@ -0,0 +1,274 @@
+use eth_types::evm_types::{Gas, GasCost, Memory, OpcodeId, ProgramCounter, Stack, Storage};
+use eth_types::{
+    fix_geth_trace_memory_size, Block, GethExecStep, GethExecTrace, Hash, Transaction, Word, H256,
+};
+use ethers_core::types::{Address, Bytes, U256, U64};
+use serde_derive::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+/// BlockResultWrapper is the payload from Scroll.
+#[derive(Deserialize, Serialize, Default, Debug)]
+pub struct BlockResultWrapper {
+    pub id: u64,
+    #[serde(rename = "blockTraces")]
+    pub block_result: BlockResult,
+}
+
+/// ZkProof is the payload to Scroll.
+#[derive(Serialize, Debug)]
+pub struct ZkProof {
+    pub id: u64,
+    #[serde(rename = "evmTranscript")]
+    pub evm_transcript: Vec<u8>,
+    #[serde(rename = "stateTranscript")]
+    pub state_transcript: Vec<u8>,
+}
+
+#[derive(Deserialize, Serialize, Default, Debug)]
+pub struct BlockResult {
+    #[serde(rename = "blockTrace")]
+    pub block_trace: BlockTrace,
+    #[serde(rename = "executionResults")]
+    pub execution_results: Vec<ExecutionResult>,
+}
+
+#[derive(Deserialize, Serialize, Default, Debug)]
+pub struct BlockTrace {
+    pub number: U64,
+    pub hash: Hash,
+    pub time: u64,
+    pub coinbase: AccountProofWrapper,
+    pub difficulty: U256,
+    pub transactions: Vec<TransactionTrace>,
+    #[serde(rename = "baseFee")]
+    pub base_fee: Option<U256>,
+    #[serde(rename = "gasLimit")]
+    pub gas_limit: u64,
+}
+
+pub type EthBlock = Block<Transaction>;
+
+impl BlockTrace {
+    pub fn to_eth_block(&self) -> EthBlock {
```

Comment:
Why not make a `From` impl instead?

Reply:
We do not need `EthBlock <-> BlockTrace` for now.  Just only `BlockTrace -> EthBlock` is enough.

Reply:
`From` does not give you a two-way implementation - it's just the idiomatic way of doing type conversions in Rust. Creating `impl From<BlockTrace> for EthBlock` will be enough.

Reply:
The `Into` impl goes the same way. From the docs:
`From<T> for U implies [Into](https://doc.rust-lang.org/std/convert/trait.Into.html)<U> for T`
So you get conversions from `T` to `U`, but not for `U` to `T`.